feat(profile): infinite scroll

LinHeMa opened the pull request:
# Changes
1. infinity scroll on profile page
2. get more functions with zod
3. using `useMemo` instead of `useEffect` for correctly render
4. GQL updates


---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1208645950390142

LinHeMa committed:
tmp

i am working on fetch more picks, bookmarks and collection for inifinity
scroll.

LinHeMa committed:
tmp

- i have to keep going to write gql and then finish the server action

LinHeMa committed:
feat(profile): add infinite scroll in picks

LinHeMa committed:
Merge branch 'dev' of https://github.com/readr-media/Sachiel into feat/infinite-scroll-at-profile

LinHeMa committed:
feat(profile): infinite scroll works for bookmarks and picks

LinHeMa committed:
fix(profile): wrong tab data

LinHeMa committed:
feat: enhance profile infinite scroll functionality

- Add GetMoreCollections query for collection data fetching
- Refactor data fetching with safer error handling using safeParse
- Optimize fetch functions using object mapping

erase2004 commented on packages/mesh/app/actions/get-more-profile-data.ts:
```diff
@@ -0,0 +1,94 @@
+) => {
+  const parseResult = getMoreMemberPicksSchema.safeParse(params)
+  if (!parseResult.success) {
+    console.error('Invalid parameters:', parseResult.error)
```
應該用 `logServerSideError` 替代 `console`

erase2004 commented on packages/mesh/app/actions/get-more-profile-data.ts:
```diff
@@ -0,0 +1,94 @@
+
+    return response?.picks?.filter((item) => item.objective === 'story') ?? []
+  } catch (error) {
+    console.error('Failed to fetch picks:', error)
```
應該用 `logServerSideError` 替代 `console`

erase2004 commented on packages/mesh/app/actions/get-more-profile-data.ts:
```diff
@@ -0,0 +1,94 @@
+  // Validate the input parameters
+  const parseResult = getMoreMemberPicksSchema.safeParse(params)
+  if (!parseResult.success) {
+    console.error('Invalid parameters:', parseResult.error)
```
應該用 `logServerSideError` 替代 `console`

erase2004 commented on packages/mesh/app/actions/get-more-profile-data.ts:
```diff
@@ -0,0 +1,94 @@
+  // Validate the input parameters
+  const parseResult = getMoreMemberPicksSchema.safeParse(params)
+  if (!parseResult.success) {
+    console.error('Invalid parameters:', parseResult.error)
```
應該用 `logServerSideError` 替代 `console`

LinHeMa replied:
同樣問題在 [fbe30fc](https://github.com/readr-media/Sachiel/pull/979/commits/fbe30fc1f221ad8f2040f71631d8948b5d6dd42f) 解決

erase2004 commented on packages/mesh/app/actions/get-more-profile-data.ts:
```diff
@@ -0,0 +1,94 @@
+import fetchGraphQL from '@/utils/fetch-graphql'
+import { getLogTraceObjectFromHeaders } from '@/utils/log'
+
+const getMoreMemberPicksSchema = z.object({
```
`getMoreMemberPicksSchema` 這應該要換個名字，畢竟不只被 `getMoreMemberPicks` 使用而已，還被 `getMoreMemberBookmarks` 和 `getMoreMemberCollections` 使用

LinHeMa replied:
感謝提醒，已重新命名於 [4e96067](https://github.com/readr-media/Sachiel/pull/979/commits/4e96067f0c2388ed981939dc72afcc938c283d69)

erase2004 commented on packages/mesh/app/actions/get-more-profile-data.ts:
`getMoreMemberPicks`、`getMoreMemberBookmarks` 和 `getMoreMemberCollections` 這三者的處理流程高度相似，可以考慮共用重複的處理流程

LinHeMa replied:
在 [dc62cd9](https://github.com/readr-media/Sachiel/pull/979/commits/dc62cd9d0bc82fddd291170eef530fed12770446) 共用

erase2004 commented on packages/mesh/graphql/query/member.gql:
`GetMorePicks` 和 `GetMoreBookmarks` 這兩個起始點是 pick 的 query，應該放在 `/query/pick.gql` 底下
`GetMoreCollections` 這個起始點是 collection 的 query，應該放在 `/query/collection.gql` 底下

LinHeMa replied:
在 [cf64037](https://github.com/readr-media/Sachiel/pull/979/commits/cf6403753ad12cd4f07a06cb6f95791190bc9da1) 處理

erase2004 commented on packages/mesh/components/story-card/story-pick-info.tsx:
```diff
@@ -41,7 +41,7 @@ export default function StoryPickInfo({
           {displayPicks?.slice(0, maxCount).map((data, index) => (
             <div
-              key={data.member?.id ?? ''}
+              key={data.member?.id + data.member.customId + index}
```
`{data.member?.id` 有 `.?` 但 `data.member.customId` 卻沒有，應該統一行為

LinHeMa replied:
[7fa01a4](https://github.com/readr-media/Sachiel/pull/979/commits/7fa01a43fc1a4e660ea1f3cee2b7ac37b051c010)

erase2004 commented on packages/mesh/app/profile/_components/article-card-list.tsx:
```diff
@@ -37,71 +57,110 @@ function ArticleCardList({
   const isCollection = items.some((item) => item.__typename === 'Collection')
 
+  const fetchMorePicksInProfile = async (pageIndex: number) => {
+    console.log('trigger', hasMoreData)
```
移除 debug 訊息

LinHeMa replied:
移除 [cc595b2](https://github.com/readr-media/Sachiel/pull/979/commits/cc595b29565be04394ebac1282a34d5d5da81e05)

erase2004 commented on packages/mesh/app/profile/_components/article-card-list.tsx:
```diff
@@ -25,8 +34,19 @@ function ArticleCardList({
+    PAGE_SIZE: 40,
+    MAX_ELEMENTS: 200,
+  } as const
+  const [hasMoreData, setHasMoreData] = useState(true)
```
如果 hasMoreData 不影響畫面 render state，可以改用 ref 來處理

LinHeMa replied:
[d209f0d](https://github.com/readr-media/Sachiel/pull/979/commits/d209f0d2daae3ec5f52c6b0c3a28ee64b238c781) 改用 `useRef` 處理

erase2004 commented on packages/mesh/app/profile/_components/article-card-list.tsx:
```diff
@@ -37,71 +57,110 @@ function ArticleCardList({
+    console.log('trigger', hasMoreData)
+    if (!hasMoreData) return []
+
+    const FETCH_FUNCTIONS = {
```
可以把這個變數放到 fetchMorePicksInProfile 外

LinHeMa replied:
[a367f0e](https://github.com/readr-media/Sachiel/pull/979/commits/a367f0e332b9fd4a65e3edf87c888d45a0dc597c) 處理

LinHeMa committed:
refactor(actions): turn console to logServerSideError

LinHeMa committed:
refactor(actions): renaming zod schema

- turn getMoreMemberPicksSchema into getMoreMemberDataArgsSchema.

LinHeMa committed:
refactor(actions): recompose the getmore functions

- use getMoreMemberData as base

LinHeMa committed:
refactor(gql): move query to proper location

- GetMorePicks, GetMoreBookmarks to `query/pick.gql`
- GetMoreCollections to `query/collection.gql`

LinHeMa committed:
refactor(story-card): unify the data access method

LinHeMa committed:
trivial(profile): remove console log

LinHeMa committed:
Refactor: Use useRef for hasMoreData

- Replaced useState with useRef for managing hasMoreData state.
- Updated the logic to use hasMoreData.current for checking and updating the state.
- This change prevents unnecessary re-renders and optimizes performance.

LinHeMa committed:
Refactor: Move FETCH_FUNCTIONS outside of fetchMorePicksInProfile

- Moved FETCH_FUNCTIONS constant outside of the fetchMorePicksInProfile function to avoid re-declaration on each function call.
- Simplified the type assertion for accessing FETCH_FUNCTIONS using keyof.
- This change improves code efficiency and readability by reducing redundant declarations.

erase2004 commented on packages/mesh/graphql/query/collection.gql:
```diff
@@ -160,3 +160,25 @@ query GetMemberCollections($creatorId: ID) {
   }
 }
+
+query GetMoreCollections($customId: String, $takes: Int = 20, $start: Int = 1) {
```
剛注意到這個起頭也是 pick，所以應放到 `query/pick.gql` 底下

LinHeMa replied:
在[13491bd](https://github.com/readr-media/Sachiel/pull/979/commits/13491bd4a3208abca3caa4b84899391b81268de0)處理

erase2004 commented on packages/mesh/app/actions/get-more-profile-data.ts:
```diff
@@ -7,88 +7,80 @@ import {
+export const getMoreMemberPicks = async (
+  params: z.infer<typeof getMoreMemberDataArgsSchema>
+) => {
+  return getMoreMemberData(
```
`getMoreMemberPicks` 原先有 `response?.picks?.filter((item) => item.objective === 'story') ?? []` filter 的行為，但在這裡是沒有的，這個行為需要補上？

LinHeMa replied:
[7e0d817](https://github.com/readr-media/Sachiel/pull/979/commits/7e0d817d8cbf849892d32a354230f2d40e6eeb0b)處理，感謝提醒

caesarWHLee commented on packages/mesh/app/profile/_components/article-card-list.tsx:
```diff
@@ -37,71 +65,102 @@ function ArticleCardList({
   const isCollection = items.some((item) => item.__typename === 'Collection')
 
+  const fetchMorePicksInProfile = async (pageIndex: number) => {
+    if (!hasMoreData) return []
```
應該是看 hasMoreData.current

LinHeMa replied:
[52415a8](https://github.com/readr-media/Sachiel/pull/979/commits/52415a8bde9a10eb6cdbe6f74bd5662912173dd0) 處理上個問題，以及這個，感謝提醒

caesarWHLee commented on packages/mesh/app/profile/_components/article-card-list.tsx:
```diff
@@ -25,8 +40,21 @@ function ArticleCardList({
+  const PAGINATION_CONFIG = {
+    PAGE_SIZE: 40,
+    MAX_ELEMENTS: 200,
+  } as const
```
const 應該放 component 外就好，同 L29

LinHeMa replied:
[3ab32ee](https://github.com/readr-media/Sachiel/pull/979/commits/3ab32ee42a9d78cf212f32a0e9ee723c8e786cd0) 處理

caesarWHLee reviewed (commented):
目前 loadmore 會在 `article-card-list` 中管理狀態，不過初始狀態是來自於 `profile-page`，會導致切換 tab 後原本 loadmore 的資料被清掉要重新拿，建議改成同一個 state 去更新以避免這個狀況

LinHeMa committed:
Merge branch 'dev' of https://github.com/readr-media/Sachiel into feat/infinite-scroll-at-profile

LinHeMa committed:
Merge branch 'feat/infinite-scroll-at-profile' of github.com:LinHeMa/Sachiel into feat/infinite-scroll-at-profile

LinHeMa committed:
fix(gql): move GetMoreCollections to pick

LinHeMa committed:
fix(actions): add back filter function

LinHeMa committed:
fix(profile): move hasMore variable to profile page

LinHeMa committed:
refactor(profile): move constants

LinHeMa committed:
Merge branch 'dev' of https://github.com/readr-media/Sachiel into feat/infinite-scroll-at-profile

erase2004 reviewed (approved):
LGTM 👏

caesarWHLee reviewed (approved):
LGTM
